lib/jsdom/vm-shim.js

```diff
@@ -62,6 +153,52 @@ exports.runInContext = function (code, contextifiedSandbox, options) {
   // make sure we keep comments
   escodegen.attachComments(ast, comments, tokens);
 
+  // TODO: these global propertise can't be delete. (configurable: false)
+  // TODO: these global propertise are defined at start of program.
+  // TODO: hoist of function declaration
+  const globalDeclarations = findGlobalDeclarations(ast);
+  for (const globalDeclaration of globalDeclarations) {
+    if (globalDeclaration.type === "VariableDeclaration") {
+      const globals = Object.keys(globalDeclaration.globals);
+      const expressions = globals.map(getPatchNode).concat({ type: "Identifier", name: globals[0] });
+      globalDeclaration.declarations.push({
+        type: "VariableDeclarator",
+        id: { type: "Identifier", name: globals[0] },
+        init: { type: "SequenceExpression", expressions }
+      });
+    } else if (globalDeclaration.type === "FunctionDeclaration") {
+      const { parent } = globalDeclaration;
+      switch (parent.type) {
+        case "Program":
+        case "BlockStatement":
+          parent.body.splice(
+            parent.body.indexOf(globalDeclaration) + 1,
+            0,
+            getPatchExpressionStatement(globalDeclaration)
+          );
+          break;
+        case "IfStatement":
+          parent[parent.consequent === globalDeclaration ? "consequent" : "alternate"] =
+            getPatchBlockStatement(globalDeclaration);
+          break;
+        case "SwitchStatement":
+          for (const { consequent } of parent.cases) {
+            const index = consequent.indexOf(globalDeclaration);
+            if (index >= 0) {
+              consequent.splice(index + 1, 0, getPatchExpressionStatement(globalDeclaration));
+              break;
+            }
+          }
+          break;
+        case "LabeledStatement":
+          parent.body = getPatchBlockStatement(globalDeclaration);
+          break;
+        default:
+          throw new Error("don't know this parent type: " + parent.type);
+      }
+    }
+  }
+
```

Comment:
The majority of what's added in this file is not tested. I was able to delete lines 180-197 with no tests failing, as well as lines 26-45 and 64-76.

It's also not clear what this code is trying to *do*. We already use https://github.com/ForbesLindesay/acorn-globals to detect globals and convert them into window assignments. I can see that it makes your new test pass, but why isn't acorn-globals good enough? Is there a bug in acorn-globals we should fix instead?

In fact, a lot of this code looks like it might be copied from acorn-globals?? What is going on??

Reply:
Yes, I addded test cases that the deletions should fail. 

findGlobals only finds all accessing of non-declared identifiers as accessing global variables, but "var" decarations and function declarations both out of all function scopes install new properties into global object. I don't ensure that it is a bug or feature of acorn-globals. The implemention of this vm-shim.js is to enclose the code into function scope, so all global declarations become local declarations, and then these new installation are dismissed. The patch is assignment being used to install these variables into global object.

I splited from my code in vm-shim.js that is modified from acorn-globals to create find-global-declarations.js and add license.

Reply:
I don't quite understand. From what I can see of the existing vm-shim code, it runs `findGlobals` on `ast`, which is created by parsing `code`, which is the raw code itself. Nothing is enclosed in a function scope, that I can see. So it should work as-is.

Probably I am missing something, but I'd appreciate your help understanding before I merge this. Maybe there is a cleaner solution that doesn't involve so much duplicate work.

Reply:
line 172:
`  return Function("window", rewrittenCode + suffix).bind(contextifiedSandbox)(contextifiedSandbox);`
The function `Function` create a function that body is `rewrittenCode + suffix`, so the code is in a function scope.

Reply:
Right, but that is the rewritten code. findGlobals is run on the original code, so it should find all var declarations.

Reply:
Right, but global declarations need differently rewriting. In addition, there is a bug that globally declared identifiers should also be found by findGlobals.

Reply:
I don't understand your first sentence.

If there are bugs in the findGlobals npm package, can you submit a pull request to them to fix them, instead of duplicating the code with a bugfix into jsdom?

Reply:
There are two bugs. I only partly fixed the bug about global declarations, but I don't fix the bug about globally declared identifiers yet.

Reply:
> If there are bugs in the findGlobals npm package, can you submit a pull request to them to fix them, instead of duplicating the code with a bugfix into jsdom?
I removed the code about the bugfix of findGlobals npm package, and then I submitted a pull request to findGlobals package: https://github.com/ForbesLindesay/acorn-globals/pull/63